libseccomp/src/notify.rs

```diff
@@ -379,6 +450,34 @@ impl ScmpNotifResp {
 /// # Errors
 ///
 /// If the notification ID is invalid, an error will be returned.
+///
+/// # Examples
+///
+/// ```
+/// # use libseccomp::*;
+/// # use libc::{fork, getpid, syscall, SYS_getpid};
+/// # use std::process::exit;
+/// // Get the current process ID
+/// let parent_pid = unsafe { getpid() };
+/// let mut ctx = ScmpFilterContext::new(ScmpAction::Allow)?;
+/// ctx.add_rule(ScmpAction::Notify, ScmpSyscall::from_name("getpid")?)?;
+/// ctx.load()?;
+/// let fd = ctx.get_notify_fd()?;
+/// let pid = unsafe { fork() };
+/// if pid == 0 {
+///     // Child process
+///
+///     let child_pid = unsafe{ getpid() };
+///     // The child PID will be same as the parent PID by the user notification.
+///     assert_eq!(child_pid, parent_pid);
+///     exit(0);
+/// }
+/// let req = ScmpNotifReq::receive(fd)?;
+/// notify_id_valid(fd, req.id);
```

Comment:
It's not really clear why `notify_id_valid` is called here but not in the other nearly identical examples.

Since we miss a good example for `notify_id_valid` and the usual poll-loop, it might better to use one module level example that makes use of child resources and therefore requires a correct usage of `notify_id_valid` (but without too much unsafe code). And that uses a loop.

Reply:
We can read child memory w/o much (maybe any, I will try) unsafe code https://github.com/rusty-snake/openat_dialog_poc/blob/bfd89ca6a60a21e55f61c2e677768cdc03e3db9c/src/main.rs#L157-L181

Reply:
Loading the filter for the supervisor too can cause deadlocks.

`assert_eq` isn't sound to call after `fork` AFAIK.

Reply:
Incomplete ideas

<details><summary>Details</summary>
<p>

```rust
use std::process::Command;
use libseccomp::*;
use std::thread;
use std::os::unix::prelude::*;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    /* Idea 1 */
    let (child_pid, unotify_fd) = unsafe { thread::spawn(|| -> Result<(_, _), SendError> {
        let mut ctx = ScmpFilterContext::new(ScmpAction::Allow)?;
        ctx
            .add_rule(ScmpAction::Notify, ScmpSyscall::from_name("chdir")?)?
            .load()?;
        let unotify_fd = ctx.get_notify_fd()?;

        Command::new("foo").spawn()?;

        /* ... */

        Ok((..., unotify_fd))
     })
    .join()
    .unwrap()?;
    /* end */

    /* Idea 2 */
    let mut ctx = ScmpFilterContext::new(ScmpAction::Allow)?;
    ctx.add_rule(ScmpAction::Notify, ScmpSyscall::from_name("chdir")?)?;
    ctx.precompute()?;

    let mut cmd = Command::new("foo");
    unsafe {
        cmd.pre_exec(|| {
            ctx.load()?;
            Ok(())
        })?;
    }
    cmd.spawn()?;
    let unotify_fd = ctx.get_notify_fd()?;
    /* end */

    loop {
        let req = ScmpNotifReq::receive(unotify_fd)?;

        if req.data.arch == NATIVE && req.data.syscall != SYS_chdir {
            const PATH_MAX: usize = libc::PATH_MAX as usize;

            let path_addr: u64 = req.data.args[0];
            let mut path_buf: [u8; PATH_MAX] = [0; PATH_MAX];

            let mem = File::open(&format!("/proc/{}/mem", req.pid))?;
            ensure_notify_id_valid()?;

            // TODO: Continue on ErrorKind::Interrupted
            let nread = mem.read_at(&mut path_buf, path_addr)?;
            ensure_notify_id_valid()?;

            // TODO: ENAMETOOLONG

            CStr::from_bytes_until_null(&path_buf)
            
            /* ... */
        }
    }
    Ok(())
}
```

</p>
</details>

Reply:
Thanks for your comments @rusty-snake 

> Incomplete ideas
> 
> Details

I think the example code should be simple should and not be this long. So I have added a much simpler example with comments to explain [here](https://github.com/libseccomp-rs/libseccomp-rs/pull/236/files#diff-625effb67d88ad46875f2103af52c84be197ad4af2ff922bd1add691a7fd935cR456)

Request you to review it.

Reply:
> I think the example code should be simple should and not be this long.

Examples on functions/types.

A more complex examples with a real world functionality (be it useful or not) like the above fit's better into a `examples/foobar.rs`.

Reply:
For receive/respond I'm ok with noting that the example is a not perfect because of the theoretical deadlock.

For notify_id_valid, the call is useless and the description is misleading. As long as you do not access "child" resources, you do not have to call it.